### PR TITLE
Add ropc command and ropc flow to BaseController

### PR DESCRIPTION
**Pull request by shahzaibj**
Related ADAL PR: https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/1638
Related MSAL PR: https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/1496

Also added E2E tests at the controller level to test ropc flow.

### CODE REVIEW

File: common4j/src/main/com/microsoft/identity/common/java/controllers/BaseController.java [hunk 1]
```diff
@@ -147,6 +150,99 @@ public abstract AuthorizationResult deviceCodeFlowAuthRequest(final DeviceCodeFl
...
+                .build();
+
+        logParameters(TAG, parametersWithScopes);
+
+        //0) Get known authority result
+        parametersWithScopes.getPlatformComponents()
+                .getPlatformUtil()
+                .throwIfNetworkNotAvailable(parametersWithScopes.isPowerOptCheckEnabled());
+
+        Authority.KnownAuthorityResult authorityResult = Authority.getKnownAuthorityResult(parametersWithScopes.getAuthority());
```
Comment (rpdome) [inline common4j/src/main/com/microsoft/identity/common/java/controllers/BaseController.java:181]: nit : final

File: common4j/src/main/com/microsoft/identity/common/java/providers/microsoft/microsoftsts/MicrosoftStsOAuth2Strategy.java [hunk 2]
```diff
@@ -430,6 +431,41 @@ public MicrosoftStsTokenRequest createRefreshTokenRequest(@NonNull final Abstrac
...
+
+        request.setUsername(parameters.getUsername());
+        request.setPassword(parameters.getPassword());
+        request.setClaims(parameters.getClaimsRequestJson());
+        request.setClientId(parameters.getClientId());
+        request.setRedirectUri(parameters.getRedirectUri());
+        request.setScope(StringUtil.join(" ", parameters.getScopes()));
+        setTokenRequestCorrelationId(request);
+
+        if (PopAuthenticationSchemeInternal.SCHEME_POP.equals(parameters.getAuthenticationScheme().getName())) {
```
Comment (rpdome) [inline common4j/src/main/com/microsoft/identity/common/java/providers/microsoft/microsoftsts/MicrosoftStsOAuth2Strategy.java:453]: Is this supported by ROPC?

**Review by rpdome — APPROVED** [review]

File: common4j/src/main/com/microsoft/identity/common/java/controllers/BaseController.java [hunk 3]
```diff
@@ -152,6 +155,99 @@ public abstract AuthorizationResult deviceCodeFlowAuthRequest(final DeviceCodeFl
...
+
+        if (tokenResult != null && tokenResult.getSuccess()) {
+            // Suppressing unchecked warnings due to casting of rawtypes to generic types of OAuth2TokenCache's instance tokenCache while calling method saveAndLoadAggregatedAccountData
+            @SuppressWarnings(WarningType.unchecked_warning) final List<ICacheRecord> savedRecords = tokenCache.saveAndLoadAggregatedAccountData(
+                    oAuth2Strategy,
+                    getAuthorizationRequest(oAuth2Strategy, parameters),
+                    tokenResult.getTokenResponse()
+            );
+
+            final ICacheRecord savedRecord = savedRecords.get(0);
```
Comment (AdamBJohnsonx) [inline common4j/src/main/com/microsoft/identity/common/java/controllers/BaseController.java:223]: check length first, or where will the NoSuchElementException go?

File: common4j/src/main/com/microsoft/identity/common/java/controllers/BaseController.java [hunk 4]
```diff
@@ -152,6 +155,99 @@ public abstract AuthorizationResult deviceCodeFlowAuthRequest(final DeviceCodeFl
...
+                .platformComponents(parameters.getPlatformComponents())
+                .build();
+
+        //1) Get oAuth2Strategy for Authority Type
+        @SuppressWarnings(WarningType.rawtype_warning) final OAuth2Strategy oAuth2Strategy = parametersWithScopes
+                .getAuthority()
+                .createOAuth2Strategy(strategyParameters);
+
+        //2) Get the token by exchanging password
+        final TokenRequest ropcTokenRequest = oAuth2Strategy.createRopcTokenRequest(
```
Comment (AdamBJohnsonx) [inline common4j/src/main/com/microsoft/identity/common/java/controllers/BaseController.java:205]: Should we check this for a null return?

File: common4j/src/main/com/microsoft/identity/common/java/commands/RopcTokenCommand.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+import com.microsoft.identity.common.java.commands.parameters.RopcTokenCommandParameters;
+import com.microsoft.identity.common.java.controllers.BaseController;
+import com.microsoft.identity.common.java.logging.Logger;
+import com.microsoft.identity.common.java.result.AcquireTokenResult;
+
+import java.util.List;
+
+import lombok.NonNull;
+
+public class RopcTokenCommand extends TokenCommand {
```
Comment (AdamBJohnsonx) [inline common4j/src/main/com/microsoft/identity/common/java/commands/RopcTokenCommand.java:36]: Maybe sprinkle some links in here to what this does...
```suggestion
/**
 * ROPC command for Resource Owner Password Credentials.
 * see: https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth-ropc
 */
public class RopcTokenCommand extends TokenCommand {
```

File: common4j/src/main/com/microsoft/identity/common/java/providers/microsoft/microsoftsts/MicrosoftStsOAuth2Strategy.java (hunk 2 above)
Comment (AdamBJohnsonx) [reply common4j/src/main/com/microsoft/identity/common/java/providers/microsoft/microsoftsts/MicrosoftStsOAuth2Strategy.java:473]: ... not according to the documents - "Always set to BEARER". You should probably fail if POP is returned, or at least make sure we have telemetry for the fact that it is happening.

**Review by AdamBJohnsonx — APPROVED** [review]
Mostly minor comments.  It looks like POP support is not public... so we should be cautious of any such system claiming support.  At least log something, better push telemetry, but maybe failure is correct.

File: common4j/src/main/com/microsoft/identity/common/java/providers/microsoft/microsoftsts/MicrosoftStsOAuth2Strategy.java (hunk 2 above)
Comment (shahzaibj) [reply common4j/src/main/com/microsoft/identity/common/java/providers/microsoft/microsoftsts/MicrosoftStsOAuth2Strategy.java:473]: Which documents are these? Can you link them here?

**Comment by AdamBJohnsonx** [comment]
I did, in a separate comment.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

File: common4j/src/main/com/microsoft/identity/common/java/controllers/BaseController.java (hunk 1 above)
Comment (shahzaibj) [reply common4j/src/main/com/microsoft/identity/common/java/controllers/BaseController.java]: Addressed

File: common4j/src/main/com/microsoft/identity/common/java/providers/microsoft/microsoftsts/MicrosoftStsOAuth2Strategy.java (hunk 2 above)
Comment (shahzaibj) [reply common4j/src/main/com/microsoft/identity/common/java/providers/microsoft/microsoftsts/MicrosoftStsOAuth2Strategy.java:455]: Addressed. Removed PoP support for now from code and indicating that even Bearer support is for testing purposes.

File: common4j/src/main/com/microsoft/identity/common/java/controllers/BaseController.java (hunk 3 above)
Comment (shahzaibj) [reply common4j/src/main/com/microsoft/identity/common/java/controllers/BaseController.java:223]: probably caught by CommandDispatcher...same way it is happen in acquireTokenInteractive...it's the same code.

**Pull request merged**